# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

### 1. Unit of Analysis + Time Window

* **Unit of Analysis (The Grain):** One single row represents the daily performance metrics for one unique web page (`content_hash_id`) belonging to one unique client (`client_hash_id`) on one specific day (`report_date`).
* **Time Window:** The dataset spans exactly one month of historical tracking data, starting on **2026-06-01** and ending on **2026-06-30**.
* **Data Provenance:** The tracking metrics originate directly from anonymized integrations with Google Search Console and Google Analytics 4, hosted securely within the FlyRank data warehouse.

In [17]:
import duckdb

# 1. Start your database helper
con = duckdb.connect()

# 2. Tell DuckDB to use your exact token securely
con.execute("INSTALL httpfs; LOAD httpfs;")
con.execute("CREATE SECRET hf_token (TYPE HUGGINGFACE, TOKEN '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');")

# 3. Pull the row counts and dates from FlyRank's sample table
query = """
SELECT
    COUNT(*) as total_rows,
    MIN(report_date) as start_date,
    MAX(report_date) as end_date
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet'
"""

print("Running your Week 3 data contract validation...")
print(con.execute(query).df())

Running your Week 3 data contract validation...


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

   total_rows start_date   end_date
0    11694072 2026-06-01 2026-06-30


## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

We are sorting our spreadsheet columns into 4 clear fields based on the database schema:

1. **Features (The Clues):** `gsc_clicks`, `gsc_impressions`, `gsc_avg_position`, `ga4_pageviews`, `sessions_organic`. (The computer uses these performance trends to learn).
2. **Label (The Goal):** To be defined/joined. (Our final target will measure if traffic is declining).
3. **Context (Human Info):** `content_hash_id`, `client_hash_id`, `report_date`. (The computer ignores these, but humans use them to know which specific page to fix).
4. **Excluded (Flags):** `client_has_gsc`, `client_has_ga4`. (We exclude these helper flags from the model features because they just indicate data source setup, not page performance).

In [18]:
# Proving these exact columns exist by pulling a tiny 5-row sample
fields_query = """
SELECT
    client_hash_id,
    content_hash_id,
    report_date,
    gsc_clicks,
    gsc_impressions,
    gsc_avg_position,
    ga4_pageviews,
    sessions_organic
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet'
LIMIT 5
"""

print("Proving our contracted fields exist in the warehouse:")
print(con.execute(fields_query).df())

Proving our contracted fields exist in the warehouse:
            client_hash_id           content_hash_id report_date  gsc_clicks  \
0  client_3ffa76342f366962  content_1a6296faee432dae  2026-06-01           0   
1  client_3ffa76342f366962  content_73f21e612565035a  2026-06-01           0   
2  client_3ffa76342f366962  content_5a5be514ff559598  2026-06-01           0   
3  client_3ffa76342f366962  content_05b377d0c8a5cfd8  2026-06-01           0   
4  client_3ffa76342f366962  content_dc34c661d63e55a9  2026-06-01           0   

   gsc_impressions  gsc_avg_position  ga4_pageviews  sessions_organic  
0                0               NaN              0                 0  
1                0               NaN              0                 0  
2                0               NaN              0                 0  
3                0               NaN              0                 0  
4                0               NaN              0                 0  


## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

### 3. Verification Assertions & Discovered Gotchas

* **Grain Validation:** Running our baseline grain query revealed an unexpected technical anomaly: certain pages show a duplicate count of 2 on the exact same `report_date` with identical `month` scopes and zeroed metrics.
* **Resolution:** The true unique row volume is **11,687,682 rows** instead of the raw 11,694,072 rows. Our modeling workflow must enforce a `DISTINCT` aggregation layer to maintain directional accuracy.
* **Missing Data Baseline:** An investigation into `gsc_avg_position = 0` establishes that approximately **0.60%** of the dataset contains missing positions, which is within acceptable tolerances for decision-support modeling.

In [19]:
# 1. Grain Verification Query
grain_query = """
SELECT
    report_date,
    client_hash_id,
    content_hash_id,
    COUNT(*) as duplicate_count
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet'
GROUP BY report_date, client_hash_id, content_hash_id
HAVING COUNT(*) > 1
LIMIT 5
"""

# 2. Missing Position Data Query
missing_query = """
SELECT
    COUNT(*) as total_rows,
    SUM(CASE WHEN gsc_avg_position = 0 THEN 1 ELSE 0 END) as blank_position_rows,
    (SUM(CASE WHEN gsc_avg_position = 0 THEN 1.0 ELSE 0 END) / COUNT(*)) * 100 as percent_missing
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet'
"""

print("--- Running Grain Test (Should be Empty) ---")
print(con.execute(grain_query).df())

print("\n--- Running Missing Data Test ---")
print(con.execute(missing_query).df())

--- Running Grain Test (Should be Empty) ---


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

  report_date           client_hash_id           content_hash_id  \
0  2026-06-13  client_e00b29e582949543  content_762f369b413f1869   
1  2026-06-13  client_e00b29e582949543  content_4b1e89b995d62ac2   
2  2026-06-13  client_e00b29e582949543  content_8e51f9165149270e   
3  2026-06-13  client_e00b29e582949543  content_81c1fc0cd0d023b9   
4  2026-06-13  client_e00b29e582949543  content_ac3831ae8497a17e   

   duplicate_count  
0                2  
1                2  
2                2  
3                2  
4                2  

--- Running Missing Data Test ---


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

   total_rows  blank_position_rows  percent_missing
0    11694072              70243.0         0.600672


In [20]:
# Let's look at one specific duplicate page to see why it has two rows
check_dup_query = """
SELECT
    report_date,
    client_hash_id,
    content_hash_id,
    month,
    gsc_clicks,
    gsc_impressions
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet'
WHERE report_date = '2026-06-14'
  AND content_hash_id = 'content_6a09fb1e96abded6'
"""
print(con.execute(check_dup_query).df())

  report_date           client_hash_id           content_hash_id    month  \
0  2026-06-14  client_def0955f7a377868  content_6a09fb1e96abded6  2026-06   
1  2026-06-14  client_def0955f7a377868  content_6a09fb1e96abded6  2026-06   

   gsc_clicks  gsc_impressions  
0           0                0  
1           0                0  


In [21]:
# This query applies our de-duplication fix to count the TRUE unique rows
clean_grain_query = """
SELECT COUNT(*) FROM (
    SELECT DISTINCT report_date, client_hash_id, content_hash_id, gsc_clicks, gsc_impressions, gsc_avg_position
    FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet'
)
"""

print("Total rows before cleaning: 11,694,072")
print("Total unique rows after cleaning:")
print(con.execute(clean_grain_query).df())

Total rows before cleaning: 11,694,072
Total unique rows after cleaning:


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

   count_star()
0      11687682


## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

In [22]:
import duckdb

# 1. Re-authenticate securely
con = duckdb.connect()
con.execute("INSTALL httpfs; LOAD httpfs;")
con.execute("CREATE SECRET hf_token (TYPE HUGGINGFACE, TOKEN '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');")

# 2. Optimized Cohort Rule: Include rows that have EITHER active search or analytics tracking
optimized_query = """
SELECT
    COUNT(*) as raw_row_count,
    SUM(CASE WHEN gsc_data_available = true OR ga4_data_available = true THEN 1 ELSE 0 END) as optimized_cohort_rows,
    SUM(CASE WHEN gsc_data_available = true AND ga4_data_available = true THEN 1 ELSE 0 END) as strict_both_rows
FROM 'hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet'
"""

print("--- Running Optimized Contract Validation ---")
df_opt = con.execute(optimized_query).df()

raw = df_opt['raw_row_count'][0]
opt_filtered = df_opt['optimized_cohort_rows'][0]

print(f"Optimized Cohort Retention: {opt_filtered:,} out of {raw:,} rows fit the contract ({(opt_filtered/raw)*100:.2f}%).")

--- Running Optimized Contract Validation ---


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Optimized Cohort Retention: 3,969,447.0 out of 11,694,072 rows fit the contract (33.94%).


### Section 4: Data Limits & Timeline Validation Results

Our final warehouse validation queries yielded the following structural boundaries:
* **Timeline Window:** 2026-06-01 to 2026-06-30 (1 full month of daily tracking sample data).
* **Final Cohort Definition:** Selected rows where EITHER Google Search Console or Google Analytics 4 data is actively available (`gsc_data_available = true OR ga4_data_available = true`).
* **Final Volume:** 3,969,447 rows meet the contract criteria ($33.94\%$ retention), providing a clean and viable data footprint for model training.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.